src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -1453,42 +1453,110 @@ private MethodSymbol FindEntryPoint(CancellationToken cancellationToken, out Imm
+                        noMainFoundDiagnostics.AddRange(specificDiagnostics);
+                        return false;
                     }
 
                     if (candidate.IsGenericMethod || candidate.ContainingType.IsGenericType)
                     {
                         // a single error for partial methods:
-                        warnings.Add(ErrorCode.WRN_MainCantBeGeneric, candidate.Locations.First(), candidate);
-                        continue;
+                        noMainFoundDiagnostics.Add(ErrorCode.WRN_MainCantBeGeneric, candidate.Locations.First(), candidate);
+                        return false;
                     }
+                    return true;
+                }
 
-                    // PROTOTYPE(async-main): CheckFeatureAvailability should be called on non-async methods 
-                    // that return Task or Task<T>
-                    if (candidate.IsAsync)
+                foreach (var candidate in entryPointCandidates)
+                {
+                    var perCandidateBag = DiagnosticBag.GetInstance();
+                    var (IsCandidate, IsTaskLike) = HasEntryPointSignature(candidate, perCandidateBag);
+
+                    if (IsTaskLike)
                     {
-                        // PROTOTYPE(async-main): Get the diagnostic to point to a smaller syntax piece.
-                        // PROTOTYPE(async-main): Switch diagnostics around if the type is not Task or Task<T>
-                        CheckFeatureAvailability(candidate.GetNonNullSyntaxNode(), MessageID.IDS_FeatureAsyncMain, diagnostics);
+                        taskEntryPoints.Add((IsCandidate, candidate, perCandidateBag));
+                    }
+                    else
+                    {
+                        if (CheckValid(candidate, IsCandidate, perCandidateBag))
+                        {
+                            intOrVoidEntryPoints.Add(candidate);
+                        }
+                        perCandidateBag.Free();
                     }
+                }
+
+                var viableEntryPoints = ArrayBuilder<MethodSymbol>.GetInstance();
 
+                foreach (var candidate in intOrVoidEntryPoints)
+                {
+                    if (candidate.IsAsync)
+                    {
+                        diagnostics.Add(ErrorCode.ERR_NonTaskMainCantBeAsync, candidate.Locations.First(), candidate);
+                        continue;
+                    }
                     viableEntryPoints.Add(candidate);
                 }
 
-                if ((object)mainType == null || viableEntryPoints.Count == 0)
+                if (viableEntryPoints.Count == 0)
+                {
+                    foreach (var (IsValid, Candidate, SpecificDiagnostics) in taskEntryPoints)
+                    {
+                        if (!CheckValid(Candidate, IsValid, SpecificDiagnostics))
```

Comment:
>if (!CheckValid(Candidate, IsValid, SpecificDiagnostics)) [](start = 24, length = 57)

It looks like this condition can be inverted and simply combined with condition in the next ```if```.